numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -358,6 +358,45 @@ PyArray_CastAnyTo(PyArrayObject *out, PyArrayObject *mp)
 }
 
 
+static NPY_CASTING
+_get_cast_safety_from_castingimpl(PyArrayMethodObject *castingimpl,
+        PyArray_DTypeMeta *dtypes[2], PyArray_Descr *from, PyArray_Descr *to)
+{
+    PyArray_Descr *descrs[2] = {from, to};
+    PyArray_Descr *out_descrs[2];
+
+    NPY_CASTING casting = castingimpl->resolve_descriptors(
+            castingimpl, dtypes, descrs, out_descrs);
+    if (casting < 0) {
+        return -1;
+    }
+    /* The returned descriptors may not match, requiring a second check */
+    if (out_descrs[0] != descrs[0]) {
+        NPY_CASTING from_casting = PyArray_GetCastSafety(
+                descrs[0], out_descrs[0], NULL);
+        casting = PyArray_MinCastSafety(casting, from_casting);
+        if (casting < 0) {
+            goto finish;
+        }
+    }
+    if (descrs[1] != NULL && out_descrs[1] != descrs[1]) {
+        NPY_CASTING from_casting = PyArray_GetCastSafety(
+                descrs[1], out_descrs[1], NULL);
+        casting = PyArray_MinCastSafety(casting, from_casting);
+        if (casting < 0) {
+            goto finish;
+        }
+    }
+
+  finish:
+    Py_DECREF(out_descrs[0]);
+    Py_DECREF(out_descrs[1]);
+    /* NPY_NO_CASTING has to be used for (NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW) */
+    assert(casting != (NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW));
+    return casting;
+}
```

Comment:
The diff is a bit hard to read, but this is just a refactor, moving the chunk of code out from `PyArray_GetCastSafety`, so that I can use it from `PyArray_CheckCastSafety()` as a function that takes the cast safety as input and can thus check `castingimpl->casting` to short circuit.

(The use of that short-cut is probably fairly moderate, since it only helps if `np.can_cast(..., ...)` returns definitely `True` – at the moment short-cutting for a definitely `False` return would even be more useful, but that is mainly because of how ufuncs currently work, and that will change.)